hvac/api/auth_methods/cert.py

```diff
@@ -0,0 +1,254 @@
+        :type name: str | unicode
+        :param cacert: The value used here is for the Vault TLS Listener CA certificate, not the CA that issued the
+            client authentication certificate. This can be omitted if the CA used to issue the Vault server certificate
+            is trusted by the local system executing this command.
+        :type cacert: str | bool
+        :param cert_pem: Location of the cert.pem used to authenticate the host.
+        :tupe cert_pem: str | unicode
+        :param key_pem: Location of the public key.pem used to authenticate the host.
+        :param key_pem: str | unicode
+        :param mount_point:
+        :type mount_point:
+        :param use_token: If the returned token is stored in the client
+        :param use_token: bool
+        :return: The response of the login request.
+        :rtype: requests.Response
+        """
+        params = {'use_token': use_token}
+        if name != "":
+            params['name'] = name
+        api_path = '/v1/auth/{mount_point}/login'.format(mount_point=mount_point)
+
+        # Must have cert checking or a CA cert. This is caught lower down but harder to grok
+        if not cacert:
+            # If a cacert is not provided try to drop down to the adapter and get the cert there.
+            # If the cacert is not in the adapter already login will also.
+            if not self._adapter._kwargs.get('verify'):
+                raise self.CertificateAuthError("cacert must be True, a file_path, or valid CA Certificate.")
+            else:
+                cacert = self._adapter._kwargs.get('verify')
+        else:
+            validate_pem_format(cacert, "verify")
+        # if cert_pem is a string its ready to be used and either has the key with it or the key is provided as an arg
+        try:
+            if validate_pem_format(cert_pem, "cert_pem"):
+                tls_update = True
+        except exceptions.VaultError as e:
+            if isinstance(e, type(exceptions.ParamValidationError())):
+                tls_update = {}
+                if not (os.path.exists(cert_pem) or self._adapter._kwargs.get('cert')):
+                    raise FileNotFoundError("Can't find the certificate.")
+                try:
+                    for tls_part, value in {'cert_pem': cert_pem, 'key_pem': key_pem}:
+                        if value != "":
+                            tls_update[tls_part] = value
+                except ValueError:
+                    tls_update = True
+            else:
+                raise e
+        if tls_update:
+            self._adapter._kwargs = {
+                **self._adapter._kwargs,
+                **{'verify': cacert, 'cert': tuple([cert_pem, key_pem])} # need to define dict as cert is a tuple
+            }
+
+        return self._adapter.login(
+            url=api_path,
+            json=params,
+        )
+
+    class CertificateAuthError(Exception):
```

Comment:
Shouldn't hvac exceptions derive from `exceptions.VaultError` class?

Reply:
It does not let you subclass Exceptions in this way. It has similar patterns in other modules with module specific error messages.